### DISCUSSION TITLE
Usage of ZipSource

### DISCUSSION THREAD

**Discussion by vivektk84** [FirelyTeam/firely-net-sdk#1757] [Q&A] [answered]
I have a use case to load profiles compressed in tar/rar/.tgz formats. Does the ZipSource support read of compressed files other than .zip? Is there any way to achieve this?

**Comment by marcovisserFurore** [comment, marked as answer]
Hi @vivektk84,

Internally we use the C# class `System.IO.Compression.ZipFile` to extract the file and this class is not able to read tar/rar/.tgz formats. Therefore we have to change the `ZipSource` class to make this possible (actually the internal class `ZipCacher`). At first sight it does not look very hard to accomplish that. 

So, we will discuss this in our team and see if we can put this in our SDK. 

You can also create a TarSource class (for example) yourself that implements the `IAsyncResourceResolver` or `IResourceResolver` interface. For inspiration you can have a look at this source code of [ZipSource.cs](https://github.com/FirelyTeam/firely-net-sdk/blob/c10e1b13521b5e625583bd48643d9f68a05fac79/src/Hl7.Fhir.Specification/Specification/Source/ZipSource.cs).

**Reply by vivektk84** [reply]
Thanks @marcovisserFurore . Support for non zip formats will definitely a value addition. I was also thinking if I can store them manually on a FHIR server and later resolve them using the WebResolver. I was exploring that option as well, but not sure the FhirClient used in the WebResolver can add custom HTTP headers while making request to the FHIR server. If this is possible, I can resolve the profile urls using the FHIR server as well. But the key here is to add certain custom headers(like authentication header) to the request created using FhirClient. Does the FhirClient has this capability to add custom headers?

**Reply by marcovisserFurore** [reply]
Since version 2.0 of the SDK, you can add `HttpClientHandler` to the client. You can see an example [here](https://docs.fire.ly/projects/Firely-NET-SDK/client/request-response.html#adding-extra-headers).

Here you an example in combination with a `WebResolver`. I would advice you to put a `CachedResolver` around it, otherwise it can hurt the performance. Btw, it will cost some performance, because for every resource it has to communicate with a FHIR server.

```c#
        private class HttpMessageHandler : HttpClientHandler
        {
            protected async override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
            {
                // do something with the headers:
                request.Headers.Add("key", "value");
                return await base.SendAsync(request, cancellationToken);
            }
        }

       // calling code
        
        var handler = new HttpMessageHandler();
        var client = new FhirClient(new Uri("https://example.com/fhir/test"), null, handler);

        var resolver = new CachedResolver(new WebResolver(uri => client));
```

I have it tested yet, but something like should work.

**Reply by vivektk84** [reply]
That's a good information.i will try it out.Thanks @marcovisserFurore
